### PR TITLE
ddl: support distributed adding index for normal table

### LINKED ISSUE #43290: support normal table

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #43290

Problem Summary:

### What is changed and how it works?
This PR supports distributed adding index for normal table

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/scheduler.go [hunk 1]
```diff
@@ -41,6 +42,7 @@ type backfillSchedulerHandle struct {
 	jc          *JobContext
 	eleTypeKey  []byte
 	totalRowCnt int64
+	isPartition bool
```
Comment (tangenta) [inline ddl/scheduler.go:45]: Is this necessary? `backfillSchedulerHandle.ptbl` should be able to get the info.

File: ddl/index.go [hunk 2]
```diff
@@ -667,7 +667,7 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
 		job.SnapshotVer = 0
 		job.SchemaState = model.StateWriteReorganization
 
-		if job.MultiSchemaInfo == nil && tblInfo.GetPartitionInfo() != nil {
+		if job.MultiSchemaInfo == nil {
```
Comment (Benjamin2037) [inline ddl/index.go:670]: Could we also remove this @tangenta

File: ddl/scheduler.go [hunk 3]
```diff
@@ -18,29 +18,34 @@ import (
...
+	db            *model.DBInfo
+	index         *model.IndexInfo
+	job           *model.Job
+	bc            ingest.BackendCtx
+	ptbl          table.PhysicalTable
+	jc            *JobContext
+	eleTypeKey    []byte
+	totalRowCnt   int64
+	isPartition   bool
+	stepForImport bool
```
Comment (Benjamin2037) [inline ddl/scheduler.go:48]: Can we just treat import as a standard step, no special implement.

File: ddl/scheduler.go [hunk 4]
```diff
@@ -100,11 +102,21 @@ func NewBackfillSchedulerHandle(taskMeta []byte, d *ddl) (scheduler.Scheduler, e
 	}
 	bh.index = indexInfo
 
+	if stepForImport {
+		bh.stepForImport = true
+		return bh, nil
+	}
+
+	d.setDDLLabelForTopSQL(jobMeta.ID, jobMeta.Query)
```
Comment (Benjamin2037) [inline ddl/scheduler.go:110]: From line 107， will we miss line 110 part setting？

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: ddl/scheduler.go [hunk 5]
```diff
@@ -258,15 +296,14 @@ func (b *backfillSchedulerHandle) SplitSubtask(ctx context.Context, subtask []by
...
-	_, _, err = b.bc.Flush(b.index.ID, ingest.FlushModeForceGlobal)
-	if err != nil {
-		if common.ErrFoundDuplicateKeys.Equal(err) {
-			err = convertToKeyExistsErr(err, b.index, b.ptbl.Meta())
-		}
-		logutil.BgLogger().Error("[ddl] flush error", zap.Error(err))
-		return nil, err
+	if consumer.getResult() != nil {
+		return nil, consumer.getResult()
+	}
```
Comment (tangenta) [inline ddl/scheduler.go:301]:
```suggestion
	if err := consumer.getResult(); err != nil {
		return nil, err
	}
```

File: ddl/ingest/backend.go [hunk 6]
```diff
@@ -169,6 +184,29 @@ func (bc *litBackendCtx) Flush(indexID int64, mode FlushMode) (flushed, imported
...
+		if err != nil {
+			return true, false, err
+		}
+		logutil.BgLogger().Info("[ddl] acquire lock success")
+		defer func() {
+			err = mu.Unlock(bc.ctx)
+			if err != nil {
+				logutil.BgLogger().Warn("[ddl] release lock error", zap.Error(err))
+			}
+			logutil.BgLogger().Info("[ddl] release lock success")
```
Comment (tangenta) [inline ddl/ingest/backend.go:202]:
```suggestion
				logutil.BgLogger().Warn("[ddl] release lock error", zap.Error(err))
			}
			logutil.BgLogger().Info("[ddl] release distributed flush lock success")
```

File: ddl/ingest/backend.go [hunk 7]
```diff
@@ -169,6 +184,29 @@ func (bc *litBackendCtx) Flush(indexID int64, mode FlushMode) (flushed, imported
...
+		if err != nil {
+			return true, false, err
+		}
+		logutil.BgLogger().Info("[ddl] acquire distributed flush lock success", zap.Int64("jobID", bc.jobID))
+		defer func() {
+			err = mu.Unlock(bc.ctx)
+			if err != nil {
+				logutil.BgLogger().Warn("[ddl] release lock error", zap.Error(err))
+			}
+			logutil.BgLogger().Info("[ddl] release distributed flush lock success")
```
Comment (tangenta) [inline ddl/ingest/backend.go:202]:
```suggestion
			if err != nil {
				logutil.BgLogger().Warn("[ddl] release distributed flush lock error", zap.Error(err), zap.Int64("jobID", bc.jobID))
			} else {
				logutil.BgLogger().Info("[ddl] release distributed flush lock success", zap.Int64("jobID", bc.jobID))
			}
```

**Review by tangenta — APPROVED** [review]

**Pull request merged**